src/isomorphism.rs

```diff
@@ -28,9 +29,156 @@ use petgraph::visit::{
     EdgeRef, GetAdjacencyMatrix, IntoEdgeReferences, NodeIndexable,
 };
 use petgraph::EdgeType;
-use petgraph::{Directed, Incoming};
+use petgraph::{Directed, Incoming, Outgoing};
 
-impl<'a, Ty> NodesRemoved for &'a StableGraph<PyObject, PyObject, Ty>
+use rayon::iter::{
+    IndexedParallelIterator, IntoParallelRefIterator, ParallelIterator,
+};
+use rayon::slice::ParallelSliceMut;
+
+type Graph<Ty> = StableGraph<PyObject, PyObject, Ty>;
+
+// NOTE: assumes contiguous node ids.
+pub trait NodeSorter<Ty>
+where
+    Ty: EdgeType,
+{
+    fn sort(&self, _: &Graph<Ty>) -> Vec<usize>;
+
+    fn reorder(&self, py: Python, graph: &Graph<Ty>) -> Graph<Ty> {
+        let order = self.sort(graph);
+
+        let mut new_graph = Graph::<Ty>::default();
+        let mut id_map: Vec<usize> = vec![0; graph.node_count()];
+        for node in order {
+            let node_index = graph.from_index(node);
+            let node_data = graph.node_weight(node_index).unwrap();
+            let new_index = new_graph.add_node(node_data.clone_ref(py));
+            id_map[node] = graph.to_index(new_index);
+        }
+        for edge in graph.edge_references() {
+            let edge_w = edge.weight();
+            let p = id_map[graph.to_index(edge.source())];
+            let c = id_map[graph.to_index(edge.target())];
+            let p_index = graph.from_index(p);
+            let c_index = graph.from_index(c);
+            new_graph.add_edge(p_index, c_index, edge_w.clone_ref(py));
+        }
+
+        new_graph
+    }
+}
+
+struct Vf2ppSorter;
+
+impl<Ty> NodeSorter<Ty> for Vf2ppSorter
+where
+    Ty: EdgeType,
+{
+    fn sort(&self, graph: &Graph<Ty>) -> Vec<usize> {
+        let n = graph.node_count();
+
+        let dout: Vec<usize> = (0..n)
+            .map(|idx| {
+                graph
+                    .neighbors_directed(graph.from_index(idx), Outgoing)
+                    .count()
+            })
+            .collect();
+
+        let mut din: Vec<usize> = vec![0; n];
+        if graph.is_directed() {
+            din = (0..n)
+                .map(|idx| {
+                    graph
+                        .neighbors_directed(graph.from_index(idx), Incoming)
+                        .count()
+                })
+                .collect();
+        }
+
+        let mut conn_in: Vec<usize> = vec![0; n];
+        let mut conn_out: Vec<usize> = vec![0; n];
+
+        let mut order: Vec<usize> = Vec::with_capacity(n);
+
+        // Process BFS level
+        let mut process = |mut vd: Vec<usize>| -> Vec<usize> {
+            // repeatedly bring largest element in front.
+            for i in 0..vd.len() {
+                let (index, &item) = vd[i..]
+                    .par_iter()
```

Comment:
I'm curious did you benchmark this? I'm a bit worried that using a parallel iterator here will actually make this slower, especially as `i` approaches `vd.len()`. Parallel iterators aren't always faster, especially for smaller iterators where the overhead often is higher than the execution time. So we should be tactical about where we leverage rayon and not just use it everywhere because we can.

Reply:
@mtreinish  I support benchmarking it, but I can give some insight on why I suggest Georgios to parallelize it.

That loop looks like it could be one of the most expensive for sparse graphs. It does a procedure similar to Selection Sort, and it is quadratic on `vd.len()` (and from what I understood `vd` contains all graph nodes at a given BFS level). So it makes sense to parallelize the inner loop that is executed multiple times, because for most of the iterations it will be an expensive loop.

For the benchmark, I suggest running it against a sparse graph and comparing it. The simplest one that comes to my mind is a [Star Graph](https://en.wikipedia.org/wiki/Star_(graph_theory)).

Ideally, we should do like in all or other functions and decide if we parallelize the loop based on the vector size:

```rust
for i in 0..vd.len() {
  let (index, &item) = if vd.len() - i >= parallel_threshold {
     // parallel loop
  }
  else {
     // fall back to sequential loop
  }
}
```

Reply:
after some benchmarks on complete, star and gnp graphs, parallel iterator is a bit slower. i tried to set a `parallel_threshold` but no improvements either. so i switched back to serial iterator.